File: backend/src/main/java/com/telehealth/pro/controller/ReportController.java
package com.telehealth.pro.controller;

import com.telehealth.pro.dto.ApiResponse;
import com.telehealth.pro.entity.MedicalReport;
import com.telehealth.pro.entity.Patient;
import com.telehealth.pro.entity.User;
import com.telehealth.pro.repository.MedicalReportRepository;
import com.telehealth.pro.repository.PatientRepository;
import com.telehealth.pro.security.CustomUserDetails;
import com.telehealth.pro.service.CloudinaryService;
import com.telehealth.pro.service.UploadResult;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.security.access.prepost.PreAuthorize;
import org.springframework.security.core.annotation.AuthenticationPrincipal;
import org.springframework.transaction.annotation.Transactional;
import org.springframework.web.bind.annotation.*;
import org.springframework.web.multipart.MultipartFile;

import java.io.IOException;
import java.util.Arrays;
import java.util.List;
import java.util.Optional;

@RestController
@RequestMapping("/api/reports")
public class ReportController {

    @Autowired
    private PatientRepository patientRepository;

    @Autowired
    private MedicalReportRepository medicalReportRepository;

    @Autowired
    private CloudinaryService cloudinaryService;

    @PostMapping
    @PreAuthorize("hasRole('PATIENT')")
    @Transactional
    public ResponseEntity<?> uploadReport(
            @AuthenticationPrincipal CustomUserDetails userDetails,
            @RequestParam("report") MultipartFile file,
            @RequestParam("title") String title) {
        
        User currentUser = userDetails.getUser();

        if (file == null || file.isEmpty()) {
            return ResponseEntity.status(HttpStatus.BAD_REQUEST)
                    .body(ApiResponse.error("Please upload a PDF or image medical report."));
        }

        if (title == null || title.trim().isEmpty()) {
            return ResponseEntity.status(HttpStatus.BAD_REQUEST)
                    .body(ApiResponse.error("Report title is required."));
        }

        Optional<Patient> patientOpt = patientRepository.findByUser_Id(currentUser.getId());
        if (patientOpt.isEmpty()) {
            return ResponseEntity.status(HttpStatus.NOT_FOUND)
                    .body(ApiResponse.error("Patient profile not found."));
        }
        Patient patient = patientOpt.get();

        try {
            UploadResult uploadResult = cloudinaryService.uploadFile(file, "medical_reports");

            MedicalReport report = MedicalReport.builder()
                    .patient(patient)
                    .fileUrl(uploadResult.getUrl())
                    .title(title)
                    .build();

            report = medicalReportRepository.save(report);

            return ResponseEntity.status(HttpStatus.CREATED)
                    .body(ApiResponse.success("Medical report uploaded successfully.", report));
        } catch (IOException e) {
            return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR)
                    .body(ApiResponse.error("Failed to upload medical report."));
        }
    }

    @GetMapping
    @PreAuthorize("hasRole('PATIENT')")
    public ResponseEntity<?> listReports(@AuthenticationPrincipal CustomUserDetails userDetails) {
        User currentUser = userDetails.getUser();
        Optional<Patient> patientOpt = patientRepository.findByUser_Id(currentUser.getId());
        if (patientOpt.isEmpty()) {
            return ResponseEntity.status(HttpStatus.NOT_FOUND)
                    .body(ApiResponse.error("Patient profile not found."));
        }

        List<MedicalReport> reports = medicalReportRepository
                .findAllByPatientPatientIdOrderByCreatedAtDesc(patientOpt.get().getPatientId());
        return ResponseEntity.ok(ApiResponse.success(reports));
    }

    @GetMapping("/{id}")
    @PreAuthorize("isAuthenticated()")
    public ResponseEntity<?> getReportDetails(
            @AuthenticationPrincipal CustomUserDetails userDetails,
            @PathVariable Long id) {
        
        User currentUser = userDetails.getUser();
        Optional<MedicalReport> reportOpt = medicalReportRepository.findById(id);
        if (reportOpt.isEmpty()) {
            return ResponseEntity.status(HttpStatus.NOT_FOUND)
                    .body(ApiResponse.error("Report not found."));
        }
        MedicalReport report = reportOpt.get();

        boolean isOwner = report.getPatient().getUser().getId().equals(currentUser.getId());
        boolean isDoctorOrAdmin = Arrays.asList("doctor", "admin").contains(currentUser.getRole());

        if (!isOwner && !isDoctorOrAdmin) {
            return ResponseEntity.status(HttpStatus.FORBIDDEN)
                    .body(ApiResponse.error("Not authorized to view this medical report."));
        }

        return ResponseEntity.ok(ApiResponse.success(report));
    }

    @DeleteMapping("/{id}")
    @PreAuthorize("hasRole('PATIENT')")
    @Transactional
    public ResponseEntity<?> deleteReport(
            @AuthenticationPrincipal CustomUserDetails userDetails,
            @PathVariable Long id) {
        
        User currentUser = userDetails.getUser();
        Optional<Patient> patientOpt = patientRepository.findByUser_Id(currentUser.getId());
        if (patientOpt.isEmpty()) {
            return ResponseEntity.status(HttpStatus.NOT_FOUND)
                    .body(ApiResponse.error("Patient profile not found."));
        }
        Patient patient = patientOpt.get();

        Optional<MedicalReport> reportOpt = medicalReportRepository
                .findByReportIdAndPatientPatientId(id, patient.getPatientId());
        if (reportOpt.isEmpty()) {
            return ResponseEntity.status(HttpStatus.NOT_FOUND)
                    .body(ApiResponse.error("Report not found or not owned by you."));
        }
        MedicalReport report = reportOpt.get();

        // Extract filename from URL
        String fileUrl = report.getFileUrl();
        String publicId = fileUrl.substring(fileUrl.lastIndexOf("/") + 1);
        if (fileUrl.contains("cloudinary")) {
            String[] parts = fileUrl.split("/");
            String filename = parts[parts.length - 1];
            String folder = parts[parts.length - 2];
            publicId = folder + "/" + filename.substring(0, filename.lastIndexOf("."));
        }

        cloudinaryService.deleteFile(publicId);
        medicalReportRepository.delete(report);

        return ResponseEntity.ok(ApiResponse.builder()
                .success(true)
                .message("Medical report deleted successfully.")
                .build());
    }
}

File: backend/src/controllers/report.controller.js
const { MedicalReport, Patient, User } = require('../models');
const { uploadFile, deleteFile } = require('../services/cloudinary.service');
const fs = require('fs');

const uploadReport = async (req, res) => {
  try {
    const { title } = req.body;

    if (!req.file) {
      return res.status(400).json({ success: false, message: 'Please upload a PDF or image medical report.' });
    }

    if (!title) {
      // Clean up the staged file if title validation fails
      if (fs.existsSync(req.file.path)) {
        fs.unlinkSync(req.file.path);
      }
      return res.status(400).json({ success: false, message: 'Report title is required.' });
    }

    // 1. Fetch Patient details
    const patient = await Patient.findOne({ where: { user_id: req.user.id } });
    if (!patient) {
      if (fs.existsSync(req.file.path)) {
        fs.unlinkSync(req.file.path);
      }
      return res.status(404).json({ success: false, message: 'Patient profile not found.' });
    }

    // 2. Upload to Cloudinary / Local storage fallback
    const uploadResult = await uploadFile(req.file.path, 'medical_reports');

    // 3. Save report record
    const report = await MedicalReport.create({
      patient_id: patient.patient_id,
      file_url: uploadResult.url,
      title: title
    });

    res.status(201).json({
      success: true,
      message: 'Medical report uploaded successfully.',
      data: report
    });
  } catch (error) {
    console.error('Upload Report Error:', error);
    // Cleanup staged files if left over
    if (req.file && fs.existsSync(req.file.path)) {
      fs.unlinkSync(req.file.path);
    }
    res.status(500).json({ success: false, message: 'Failed to upload medical report.' });
  }
};

const listReports = async (req, res) => {
  try {
    const patient = await Patient.findOne({ where: { user_id: req.user.id } });
    if (!patient) {
      return res.status(404).json({ success: false, message: 'Patient profile not found.' });
    }

    const reports = await MedicalReport.findAll({
      where: { patient_id: patient.patient_id },
      order: [['created_at', 'DESC']]
    });

    res.status(200).json({ success: true, data: reports });
  } catch (error) {
    console.error('List Reports Error:', error);
    res.status(500).json({ success: false, message: 'Failed to retrieve reports.' });
  }
};

const getReportDetails = async (req, res) => {
  try {
    const { id } = req.params;

    const report = await MedicalReport.findByPk(id, {
      include: [{ model: Patient, include: [{ model: User, attributes: ['name', 'id'] }] }]
    });

    if (!report) {
      return res.status(404).json({ success: false, message: 'Report not found.' });
    }

    // Auth checks: User must be either the owner (patient) or a doctor/admin
    const userId = req.user.id;
    const isOwner = report.Patient.user_id === userId;
    const isDoctorOrAdmin = ['doctor', 'admin'].includes(req.user.role);

    if (!isOwner && !isDoctorOrAdmin) {
      return res.status(403).json({ success: false, message: 'Not authorized to view this medical report.' });
    }

    res.status(200).json({ success: true, data: report });
  } catch (error) {
    console.error('Get Report Details Error:', error);
    res.status(500).json({ success: false, message: 'Failed to retrieve report details.' });
  }
};

const deleteReport = async (req, res) => {
  try {
    const { id } = req.params;

    const patient = await Patient.findOne({ where: { user_id: req.user.id } });
    if (!patient) {
      return res.status(404).json({ success: false, message: 'Patient profile not found.' });
    }

    const report = await MedicalReport.findOne({
      where: { report_id: id, patient_id: patient.patient_id }
    });

    if (!report) {
      return res.status(404).json({ success: false, message: 'Report not found or not owned by you.' });
    }

    // Try to parse file ID for storage deletion
    // For local fallback, file name is stored. For Cloudinary, public_id might be extractable.
    let publicId = report.file_url.split('/').pop(); // fallback guess
    if (report.file_url.includes('cloudinary')) {
      const parts = report.file_url.split('/');
      const filename = parts.pop();
      const folderName = parts.pop();
      publicId = `${folderName}/${filename.split('.')[0]}`;
    }

    await deleteFile(publicId);
    await report.destroy();

    res.status(200).json({ success: true, message: 'Medical report deleted successfully.' });
  } catch (error) {
    console.error('Delete Report Error:', error);
    res.status(500).json({ success: false, message: 'Failed to delete report.' });
  }
};

module.exports = {
  uploadReport,
  listReports,
  getReportDetails,
  deleteReport
};
